src/Plugin/Action/GenerateImageDerivative.php

```diff
@@ -0,0 +1,218 @@
+    $source_field = $this->mediaSource->getSourceFieldName($source_media->bundle());
+    $files = $source_media->get($source_field)->referencedEntities();
+    $file = reset($files);
+    $data['source_uri'] = $file->url('canonical', ['absolute' => TRUE]);
+
+    $derivative_term = $this->utils->getTermForUri($this->configuration['derivative_term_uri']);
+    $route_params = ['node' => $entity->id(), 'media_type' => 'image', 'taxonomy_term' => $derivative_term->id()];
+    $data['destination_uri'] = Url::fromRoute('islandora.media_source_put_to_node', $route_params)
+      ->setAbsolute()
+      ->toString();
+
+    $parts = explode('/', $data['mimetype']);
+    if (count($parts) > 1) {
+      $extension = $parts[1];
+    } elseif (!empty($data['mimetype'])) {
+      $extension = $data['mimetype'];
+    } else {
+      $extension = "jpeg";
+    }
+
+    $parts = explode('#', $this->configuration['derivative_term_uri']);
+    if (count($parts) > 1) {
+      $name = $entity->id() . ' - ' . $parts[1];
+    } else {
+      $name = $entity->id() . ' - Derivative';
+    }
+    $data['filename'] = "$name.$extension";
+
+    unset($data['source_term_uri']);
+    unset($data['derivative_term_uri']);
+    return $data;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
+    $form = parent::buildConfigurationForm($form, $form_state);
+    $form['event']['#disabled'] = 'disabled';
+
+    $form['source_term'] = [
+      '#type' => 'entity_autocomplete',
+      '#target_type' => 'taxonomy_term',
+      '#title' => t('Source term'),
+      '#default_value' => $this->utils->getTermForUri($this->configuration['source_term_uri']),
+      '#required' => TRUE,
+      '#description' => t('Term indicating the source media'),
+    ];
+    $form['derivative_term'] = [
+      '#type' => 'entity_autocomplete',
+      '#target_type' => 'taxonomy_term',
+      '#title' => t('Derivative term'),
+      '#default_value' => $this->utils->getTermForUri($this->configuration['derivative_term_uri']),
+      '#required' => TRUE,
+      '#description' => t('Term indicating the derivative media'),
+    ];
+    $form['mimetype'] = [
+      '#type' => 'textfield',
+      '#title' => t('Mimetype'),
+      '#default_value' => $this->configuration['mimetype'],
+      '#required' => TRUE,
+      '#rows' => '8',
+      '#description' => t('Mimetype to convert to (e.g. image/jpeg, image/png, etc...)'),
+    ];
+    $form['args'] = [
+      '#type' => 'textfield',
+      '#title' => t('Additional arguments'),
+      '#default_value' => $this->configuration['args'],
+      '#rows' => '8',
+      '#description' => t('Additional command line arguments for ImageMagick convert (e.g. -resize 50%'),
+    ];
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitConfigurationForm(array &$form, FormStateInterface $form_state) {
```

Comment:
Should there be a validation that the `$form['mimetype']` doesn't have a non-image mimetype in it?

Reply:
good catch, could probably we a select box to limit the options and the errors

Reply:
that's a good idea